buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java

```diff
@@ -314,120 +305,306 @@ void reallocate(int size, int maxCapacity, AdaptiveByteBuf into) {
                 mags = magazines;
-                if (mags.length >= MAX_STRIPES || mags.length > currentLength || freed) {
-                    return true;
+                int mask = mags.length - 1;
+                int index = (int) (threadId & mask);
+                for (int i = 0, m = Integer.numberOfTrailingZeros(~mask); i < m; i++) {
+                    Magazine mag = mags[index + i & mask];
+                    if (buf == null) {
+                        buf = mag.newBuffer();
+                    }
+                    if (mag.tryAllocate(size, maxCapacity, buf, reallocate)) {
+                        // Was able to allocate.
+                        return buf;
+                    }
                 }
-                int preferredChunkSize = mags[0].sharedPrefChunkSize;
-                Magazine[] expanded = new Magazine[mags.length * 2];
-                for (int i = 0, l = expanded.length; i < l; i++) {
-                    Magazine m = new Magazine(this);
-                    m.localPrefChunkSize = preferredChunkSize;
-                    m.sharedPrefChunkSize = preferredChunkSize;
-                    expanded[i] = m;
+                expansions++;
+            } while (expansions <= EXPANSION_ATTEMPTS && tryExpandMagazines(mags.length));
+
+            // The magazines failed us; contention too high and we don't want to spend more effort expanding the array.
+            if (!reallocate && buf != null) {
+                buf.release(); // Release the previously claimed buffer before we return.
+            }
+            return null;
+        }
+
+        long usedMemory() {
+            long sum = 0;
+            for (Chunk chunk : chunkReuseQueue) {
+                sum += chunk.capacity();
+            }
+            if (threadLocalMagazine != null) {
+                sum += threadLocalMagazine.usedMemory.get();
+            } else {
+                for (Magazine magazine : magazines) {
+                    sum += magazine.usedMemory.get();
                 }
-                magazines = expanded;
-            } finally {
-                magazineExpandLock.unlockWrite(writeLock);
             }
-            for (Magazine magazine : mags) {
-                magazine.free();
+            return sum;
+        }
+
+        private boolean tryExpandMagazines(int currentLength) {
+            if (currentLength >= MAX_STRIPES) {
+                return true;
+            }
+            final Magazine[] mags;
+            long writeLock = magazineExpandLock.tryWriteLock();
+            if (writeLock != 0) {
+                try {
+                    mags = magazines;
+                    if (mags.length >= MAX_STRIPES || mags.length > currentLength || freed) {
+                        return true;
+                    }
+                    Magazine firstMagazine = mags[0];
+                    Magazine[] expanded = new Magazine[mags.length * 2];
+                    for (int i = 0, l = expanded.length; i < l; i++) {
```

Comment:
nit: 
```suggestion
                    for (int i = 0; i < expanded.length; i++) {
```

Reply:
Your proposed change will load the length field at every iteration, while the existing code loads in once.

Reply:
Shouldn't the JIT be able to optimize it ? That said its a nit and so feel free to ignore